pycloudlib/ibm/instance.py

```diff
@@ -479,35 +481,132 @@ def __init__(
         instance: dict,
-        floating_ip: dict,
         username: Optional[str] = None,
         **kwargs,
     ) -> "IBMInstance":
-        """Instantiate `self` from `instance` associated to `floating_ip`."""
-        nic_id = instance["primary_network_interface"]["id"]
-
-        ibm_instance_type = _IBMInstanceType.from_raw_instance(instance)
-        ibm_instance_type.add_instance_network_interface_floating_ip(
-            client,
-            id=floating_ip["id"],
-            instance_id=instance["id"],
-            network_interface_id=nic_id,
-        ).get_result()
+        """
+        Instantiate IBMInstance object from raw instance dict.
 
+        Args:
+            args: positional arguments
+            client: IBM VPC client
+            instance: raw instance dict
+            username: username to use for SSH
+            kwargs: other keyword arguments
+        """
         return cls(
             *args,
             client=client,
             instance=instance,
-            floating_ip=floating_ip,
             username=username,
             **kwargs,
         )
 
+    def _choose_from_existing_floating_ips(
+        self,
+        zone: str,
+        name_includes="default-floating-ip",
+    ) -> dict:
+        """
+        Choose a random floating IP from existing floating IPs via substring match.
+
+        Args:
+            name_includes: string, name to filter floating IPs by
+
+        Returns:
+            A floating IP object
+
+        Raises:
+            IBMException: If no floating IPs are found matching the substring
+            IBMCapacityException: If all floating IPs matching the substring are already in use
+        """
+        floating_ips = list(
+            _iter_resources(
+                self._client.list_floating_ips,
+                resource_name="floating_ips",
+                # Select that match the desired name
+                filter_fn=lambda ip: name_includes in ip["name"]
+                and ip.get("zone")
+                and ip["zone"]["name"] == zone,
+            )
+        )
+        if not floating_ips:
+            raise IBMException(
+                f"No floating IPs found matching substring {name_includes}"
+            )
+        # filter out floating ips that are already associated with an instance
+        floating_ips = [ip for ip in floating_ips if "target" not in ip]
+        if not floating_ips:
+            raise IBMCapacityException(
+                f"All floating IPs matching substring {name_includes}",
+                "are already in use.",
+            )
+
+        # pick random floating ip to help reduce contention
+        floating_ip = random.choice(floating_ips)
+
+        self._log.info(
+            "Using existing floating ip '%s' with address %s",
+            floating_ip["name"],
+            floating_ip["address"],
+        )
+        return floating_ip
+
+    def attach_floating_ip(self, floating_ip: dict) -> Optional[dict]:
```

Comment:
Is this method intended to be part of the public API? Or should we make it private?

Reply:
hmmm... this does not provide much utility as a public api method so will make this private as well.